pkg/bloombuild/planner/planner_test.go

```diff
@@ -319,3 +332,163 @@ func Test_blockPlansForGaps(t *testing.T) {
 		})
 	}
 }
+
+func Test_BuilderLoop(t *testing.T) {
+	const (
+		nTasks    = 100
+		nBuilders = 10
+	)
+	logger := log.NewNopLogger()
+
+	limits := &fakeLimits{}
+	cfg := Config{
+		PlanningInterval:        1 * time.Hour,
+		MaxQueuedTasksPerTenant: 10000,
+	}
+	schemaCfg := config.SchemaConfig{
+		Configs: []config.PeriodConfig{
+			{
+				From: parseDayTime("2023-09-01"),
+				IndexTables: config.IndexPeriodicTableConfig{
+					PeriodicTableConfig: config.PeriodicTableConfig{
+						Prefix: "index_",
+						Period: 24 * time.Hour,
+					},
+				},
+				IndexType:  types.TSDBType,
+				ObjectType: types.StorageTypeFileSystem,
+				Schema:     "v13",
+				RowShards:  16,
+			},
+		},
+	}
+	storageCfg := storage.Config{
+		BloomShipperConfig: bloomshipperconfig.Config{
+			WorkingDirectory:    []string{t.TempDir()},
+			DownloadParallelism: 1,
+			BlocksCache: bloomshipperconfig.BlocksCacheConfig{
+				SoftLimit: flagext.Bytes(10 << 20),
+				HardLimit: flagext.Bytes(20 << 20),
+				TTL:       time.Hour,
+			},
+		},
+		FSConfig: local.FSConfig{
+			Directory: t.TempDir(),
+		},
+	}
+
+	// Create planner
+	planner, err := New(cfg, limits, schemaCfg, storageCfg, storage.NewClientMetrics(), nil, logger, prometheus.DefaultRegisterer)
+	require.NoError(t, err)
+
+	// Start planner
+	err = planner.StartAsync(context.Background())
+	require.NoError(t, err)
+	require.Eventually(t, func() bool {
+		return planner.State() == services.Running
+	}, 15*time.Second, 10*time.Millisecond)
+	defer func() {
+		planner.StopAsync()
+		require.Eventually(t, func() bool {
+			return planner.State() != services.Running
+		}, 1*time.Minute, 100*time.Millisecond)
+	}()
```

Comment:
You can use functions from the `services` package for starting and stopping:


```go
err := services.StartAndAwaitRunning(context.Background(), planner)
require.NoError(t, err)
t.Cleanup(func() {
    err := services.StopAndAwaitTerminated(context.Background(), planner)
    require.NoError(t, err)
})
```

Reply:
Oh nice, TIL! Thank you 👍